File: Arrays/1-Two_Sum/1-Two_Sum.cpp
//Brute Force Approach
// Time Complexity - O(N^2)
//Space Complexity - O(1)

vector<int> twoSum(vector<int>& nums, int target) {
    for(int i = 0; i < nums.size(); i++){
        int numToFind = target - nums[i];
        for(int j = i+1; j < nums.size(); j++){
            if(nums[j] == numToFind) return {i, j};
        }
    }
    return {-1, 1};
}


//Better Approach
//Time Complexity - O(N), Each map lookup and insert is average-case O(1),  
//In rare worst-case (due to hash collisions), map operations can degrade to O(n) per op, but in practice, it's very fast
//Space Complexity - O(N)

vector<int> twoSum(vector<int>& nums, int target) {
    unordered_map<int, int> mp;
    for(int i = 0; i < nums.size(); i++){
       int numToFind = target - nums[i];
       if(mp.find(numToFind) != mp.end()) return {mp[numToFind], i};
       mp[nums[i]] = i;
    }
    return {0, 0};
 }

File: Arrays/1-Two_Sum/1-Two_Sum.java
//Brute Force Approach
// Time Complexity - O(N^2)
//Space Complexity - O(1)


for (int i = 0; i < nums.length; i++) {
    int numToFind = target - nums[i];
    for (int j = i + 1; j < nums.length; j++) {
        if (nums[j] == numToFind) {
            return new int[]{i, j};
        }
    }
}
return new int[]{-1, 1};


//Better Approach
//Time Complexity - O(N)
//Space Complexity - O(N)
public int[] twoSum(int[] nums, int target) {
        
    Map<Integer, Integer> mp = new HashMap<>();
    for(int i = 0; i < nums.length; i++){
      int numToFind = target - nums[i];
      if(mp.containsKey(numToFind)) return new int[] {mp.get(numToFind), i};
      mp.put(nums[i], i);
    }
    return new int[]{0, 0};
  }